### PR TITLE
[OSX] Distributed notification center

### PR DESCRIPTION
**Pull request by dylan-cooper**
Addresses #1861 and #4954.

This seemed to me like an appropriate solution to this problem, but it isn't like anything that you are doing currently in electron so if I am doing something crazy, let me know and I won't take offence :)

The basic idea is that this can be used rather than implementing an API like the one for dark mode with every system preference (like the graphite appearance setting) one by one.  This provides a mechanism for letting users create front-ends as separate npm packages like this one: https://github.com/dylan-cooper/electron-osx-prefs while electron is used as the back-end.  Note that it can also be used for things unrelated to the system preferences such as bluetooth turning on/off, new users being added, etc.

The reason that I did this in electron rather than in its own package is because you need an Objective-C runtime to listen to the notifications.  I initially tried with nodobjc, but it didn't work because interfacing the two runtime loops was non-trivial.  I assumed that this would come up if I tried to do this as a separate add-on as well, and the reason this worked was because it was using the chromium run time behind the scenes, but if that will work and this is a bad fit for the electron core, I can give that a shot.

### CODE REVIEW

File: spec/api-app-spec.js [hunk 1]
```diff
@@ -87,6 +87,30 @@ describe('app module', function() {
...
+      );
+
+      app.on('example-notification', function(){
+        done();
+      });
+
+      //Emit an event
+      ChildProcess.exec('echo \'import Foundation;' +
+        'NSDistributedNotificationCenter.defaultCenter().postNotificationName'+
+        '("ExampleNotification", object: nil);\' | swift -'
```
Comment (phamdaniel) [inline spec/api-app-spec.js]: I'm pretty sure we don't have access to the `swift` command on Travis-CI. I don't know if there would be another way of posting a notification.

Comment (dylan-cooper) [inline spec/api-app-spec.js]: Hmm, looks like it was just a timeout error, I increased the timeout value to 12000 and was able to run script/cibuild successfully on my own machine, will update the PR. :)

File: atom/browser/browser.cc [hunk 2]
```diff
@@ -185,4 +185,9 @@ void Browser::PlatformThemeChanged() {
   FOR_EACH_OBSERVER(BrowserObserver, observers_, OnPlatformThemeChanged());
 }
 
+void Browser::DistributedNotificationEvent(const std::string& eventName) {
+  FOR_EACH_OBSERVER(BrowserObserver, observers_, \
```
Comment (zcbenz) [inline atom/browser/browser.cc]: The trailing `\` is not needed in C++.

**Comment by zcbenz** [comment]
> I initially tried with nodobjc, but it didn't work because interfacing the two runtime loops was non-trivial. I assumed that this would come up if I tried to do this as a separate add-on as well, and the reason this worked was because it was using the chromium run time behind the scenes, but if that will work and this is a bad fit for the electron core, I can give that a shot.

This can be easily turned into a native node module, and there is no need to interface with runtime loops, you can find [scrollbar-style module](https://github.com/atom/scrollbar-style) as an example.

But I'm good making this as a built-in feature of Electron, and I love the idea of using `NSDistributedNotificationCenter` to make the interface flexible. However I tend to make it as a method of a new module, instead of keeping adding things to `app` module. A possible API looks like this:

```js
// In the main process.
const {systemPreferences} = require('electron')
systemPreferences.subscribeDistributedNotification('AppleAquaColorVariantChanged', function() {
  console.log('Aqua Color changed')
})
systemPreferences.unsubscribeDistributedNotification('AppleAquaColorVariantChanged')
```

And in future new APIs for #1861 can also be added to this module.

**Comment by dylan-cooper** [comment]
Thanks for the feedback and for pointing out the scrollbar-style module.  That looks like a great starting place for me.

I'll find some time over the weekend and prototype what it would look like in its own module and post it back here to see whether it makes sense to build it into electron.

**Comment by lgalabru** [comment]
@dylan-cooper any plan concerning systemPreferences.postNotification :) ?

**Pull request closed without merging**